hack/cmd/teststale/teststale.go

```diff
@@ -0,0 +1,203 @@
+	p.testImports = scanLineList(scanner, "TestImports: ")
+	p.xTestGoFiles = scanLineList(scanner, "XTestGoFiles: ")
+	p.xTestImports = scanLineList(scanner, "XTestImports: ")
+
+	if err := cmd.Wait(); err != nil {
+		return nil, fmt.Errorf("command did not complete: %v", err)
+	}
+	return p, nil
+}
+
+func (p *pkg) isStale(buildTime time.Time) bool {
+	// If the package itself is stale, then we have to rebuild the whole thing anyway.
+	if p.stale {
+		return true
+	}
+
+	// Test for file staleness
+	for _, f := range p.testGoFiles {
+		if isStale(buildTime, filepath.Join(p.dir, f)) {
+			log.Printf("test Go file %s is stale", f)
+			return true
+		}
+	}
+	for _, f := range p.xTestGoFiles {
+		if isStale(buildTime, filepath.Join(p.dir, f)) {
+			log.Printf("external test Go file %s is stale", f)
+			return true
+		}
+	}
+
+	format := "{{.Stale}}"
+	imps := []string{}
+	imps = append(imps, p.testImports...)
+	imps = append(imps, p.xTestImports...)
+
+	cmd := newCmd(format, imps)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		log.Printf("unexpected error with creating stdout pipe: %v", err)
+		return true
+	}
+	// Start executing the command
+	if err := cmd.Start(); err != nil {
+		log.Printf("unexpected error executing command: %v", err)
+		return true
+	}
+
+	// Parse the command output
+	scanner := bufio.NewScanner(stdout)
+	scanner.Split(bufio.ScanLines)
+
+	for i := 0; scanner.Scan(); i++ {
+		if out := scanner.Text(); out != "false" {
+			log.Printf("import %q is stale: %s", imps[i], out)
+			return true
+		}
+	}
+
+	if err := cmd.Wait(); err != nil {
+		log.Printf("unexpected error waiting to finish: %v", err)
+		return true
+	}
+	return false
+}
+
+// scanLineList scans a line, removes the prefix and splits the remaining line into
+// individual strings.
+// TODO: There are ton of intermediate strings being created here. Convert this to
+// a bufio.SplitFunc instead.
+func scanLineList(scanner *bufio.Scanner, prefix string) []string {
+	scanner.Scan()
+	list := strings.TrimPrefix(scanner.Text(), prefix)
+	line := strings.Trim(list, "[]")
+	if len(line) == 0 {
+		return []string{}
+	}
+	return strings.Split(line, " ")
+}
+
+func isStale(buildTime time.Time, filename string) bool {
+	stat, err := os.Stat(filename)
+	if err != nil {
+		return true
+	}
+	return stat.ModTime().After(buildTime)
+}
+
+// IsTestStale checks if the test binary is stale and needs to rebuilt.
+// Some of the ideas here are inspired by how Go does staleness checks.
+func isTestStale(binPath, pkgPath string) bool {
+	bStat, err := os.Stat(binPath)
+	if err != nil {
+		log.Printf("Couldn't obtain the modified time of the binary: %v", err)
+		return true
+	}
+	buildTime := bStat.ModTime()
+
+	p, err := newPkg(pkgPath)
+	if err != nil {
+		log.Printf("Couldn't retrieve the test package information: %v", err)
+		return false
+	}
+
+	return p.isStale(buildTime)
+}
+
+func main() {
+	flag.Parse()
+	if isTestStale(*binary, *pkgPath) {
+		fmt.Println("true")
```

Comment:
Thought about that, but I wasn't very comfortable exiting with 1 when the program itself actually succeeded. Will do that if you insist.